baybe/surrogates/gaussian_process/core.py

```diff
@@ -28,14 +37,23 @@ class GaussianProcessSurrogate(Surrogate):
     # See base class.
 
     # Object variables
-    kernel: Optional[Kernel] = field(default=None)
-    """The kernel used by the Gaussian Process."""
+    kernel_factory: KernelFactory = field(
```

Comment:
whats the reason to use factories here instead of a direct kernel?
It makes the specification via json a bit stranger

Reply:
For the JSON part, see the [other thread](https://github.com/emdgroup/baybe/pull/235#discussion_r1601175560).

For why a plain kernel does not suffice: @AVHopp already stumbled across the same problem when he added the `kernel` field to the class. The problem with directly providing a kernel is that you have to mechanism to specify the properties of your kernel in a context-specific way, since you don't have access to things like `searchspace` and training data, which will only be encountered during the `fit` routine. That is, it wouldn't even allow us to replicate the EDBO logic unless it's hard-coded inside the class body, but our goal was to make this configurable. You basically asked Alex already the same question why he didn't somehow put the defaults into the attributes. The above is the answer: it's not possible. At least I cannot see any way to do so without using a factory. But I'm open to better ideas in case you have any?

Reply:
all ok about the converter and the json thing, i see now the converter makes it easy

The only thing thats bugging me then is the variable name, while `factory` is strictly correct, as a user i want to specify a `kernel`

Reply:
Yes, agree, I'm also not 100% happy. If we want to keep the flexibility, I basically see two solutions:
1. (Current approach) Mention the implicit conversion in the docstring, i.e. that it's simply OK to pass a kernel
2. Reflect the flexibility in the constructor signature, as mentioned in the other thread. While the converted thing is ALWAYS a factory (and hence the field name should stay as such), we could expose it as "kernel_or_factory/factory_or_kernel" in the construction by setting an `alias`. That would make it crisp clear to the user, with the downside of having a more complex name in the constructor

Reply:
And follow-up to 2: plus second downside (which is a general problem though, not only related to this situation) that in our autodoc, there is currently no distinction between attributes and constructor arguments. So it wouldn't be exactly clear to me what the docstring should say. Ideally, we'd want separate docstrings in this case...

Reply:
Hey @Scienfitz, I just realized there is a third option and quickly wanted to align with which one to chose. Note that in the end we'll have 3 attributes to be managed: kernel, likelihood and mean, below indicated with `<attr>`. So here again the options:
 
1) Simply using `<attr>_factory` with converters in place that are mentioned in the docstring. Resulting in 3 parameters exposed in the signature.
2) Using fields `<attr>_factory` internally but exposing them as `<attr>_or factory`, making the conversion explicit to the user. Again resulting in 3 exposed parameters.
3) Using fields `<attr>_factory` internally but exposing two fields per attribute, namely `<attr>` and <attr>_factory`. That would result in 6 parameters in the signature, giving a strict separation in the UI. For instance, If the user provides a `Kernel`, they would need to pass it to `kernel`, and if the want to pass a `KernelFactory` they pass it to `kernel_factory`.

Reply:
hmm I'd prefer 2 but if it make issues with the doc then thats bad. Between 1 and 3 I have to strict opinion, any good args for/against any of them? otherwise feel free to choose

Further question:
I can see the need for kernel + kernel_factory, what about the other args? Dot hey really also need a factory or would just `<attr>` do it? You know overall I think the factory as the default seems required to service the different prior regimes from our side. BUT if a user decides to usse a custom kernel I think theyd rather do simpe prelim studies and then decide on one particular prior and kernel.  they would prob not design a factory, so the overall user-gain from having the fatories seems limited to me

Reply:
Hm, even though the doc stuff is not ideal, I'd still say let's go with what we think is most elegant in terms of code, and not what the docs dictate (we can always find a workaround here / fix the docs issue in the future). So if you prefer 2, then I'll go with 2.

For the other question: yes, if we want to keep the existing noise logic, we need a factory for the likelihood as well, because the noise setting depends on the context. So same line of reasoning. So what do you think?

The mean is the only one for we currently don't need it, so for we have three options for it:
1. Don't expose it at all (always implicitly using a `ConstantMean`) and only add it once we have a real use case for it
2. Add it with just `<attr>`, but that would be a bit inconsistent given that for the other two we have factory options
3. Do the full extension in the same style as for kernel and likelihood

Reply:
if the mean is something you think a user would consider configuring then lets add it here (if it doesnt require other complex additions) witht he same treatment as others, otherwise not

Reply:
Yes, I think there are enough cases where you'd want to change the mean (at least switching between the not so exotic cases zero-mean, constant-mean and linear-mean). Of course, you can also go wild here, just like for the kernel... But ok, I'll then tackle this in the immediate follow-up PR. Will now finalize this one here with the changes discussed in the open threads and then we can merge 🚀